Move source files to separate folder (take 2)

basdelfos opened the pull request:
Made a new PR for only moving the source files to a separate folder.
Gulp 4 made some changes easier and now it also doesn't need additional gulp packages for moving files around.

@McGiverGim: could you test if it builds fine on windows and linux?

McGiverGim commented:
Yeah, tested under windows and linux. Both seem to work right. Both generate the installers, and they execute without problem. 🙄

basdelfos commented:
@McGiverGim: thanks for testing. Everything looks well for this PR, I will let it rest for one day for others to see it. After that we can merge it.

mikeller commented on gulpfile.js:
```diff
@@ -200,128 +200,128 @@ function clean_cache() {
-        './tabs/setup_osd.js',
-        './tabs/transponder.js',
-        './main.js',
+        './src/js/libraries/q.js',
```
`libraries` should really go into a `lib/` folder or something, and not into the `src/` folder, since they are not sources of the project itself. The same goes for `css/fonts/`.

basdelfos replied:
@mikeller: I didn't touch the folder structure under the 'js' folder, this was already how it has been setup. As these are client libraries (running in the browser) this isn't a strange place to put them.

mikeller replied:
I know you didn't. I always thought it was wrong and confusing to have a folder that contains a mix of files that we created and files that were copied from the internet, and have a `js/` folder, and `.js` files that live outside of this folder. I never changed it since before we used `gulp` and just zipped and shipped the entire repository to the Google web store, any new structure was still going to be a compromise.
Now that we have gulp, and use a build process to assemble the the application from parts, we should really clean this up and not just move a mess from one place to another.

mikeller replied:
If you are not comfortable with doing these changes, we can just move all of the existing application files (without the changelog) into the `src/` folder for now, and do the cleanup in a later step. But I strongly believe that the cleanup should be done.

basdelfos replied:
Yes, no problem, now we have the hood open, just as good do it the write way.
Any other files you want to move around?

mikeller replied:
I think that's it for now - if we find stuff later that we are not happy with we can move it then. Doing it now will also give us some time for bugs to be found if the pathing is not correct after the move, or we forgot something.

mikeller commented on gulpfile.js:
```diff
@@ -200,128 +200,128 @@ function clean_cache() {
+        './src/*.html',
+        './src/tabs/*.html',
+        './src/images/**/*',
+        './src/_locales/**/*',
```
I think `_locales/` is another case that should be outside of the `src/` folder.

basdelfos replied:
Agreed, after a better thought - these are files that are maintained by other people (and maybe do not have the knowledge of nw.js apps) it is more logical to place this folder in the root.

mikeller replied:
The `_locales/` folder? I agree on that - they pretty much a type of file on its own.

basdelfos replied:
I will also rename it to something like `./translations` and let gulp copy the files to `./dist/_locales`.

basdelfos commented:
@mikeller:

> As I said before, changelog.html has to stay in the root folder (or we have to introduce a copy of the changelog in the root folder, which will require additional work whenever the changelog is updated).

OK, no problem for me.

> Also, is osx/dmg-background@2x.png following a fixed naming convention? The @ in the file name has the potential to cause problems for some file systems, so if it's not a fixed naming convention, it should be avoided.

Yes, this is a fixed naming convention, when the app (or in this case the installer) is runs on a Retina screen, macOS automatically uses the @2x file to display the high-res image. When the @2x file doesn't exist it takes the low-res image and scales it 2x which results in a blurry image.

mikeller commented:
> Yes, this is a fixed naming convention, when the app (or in this case the installer) is runs on a Retina screen, macOS automatically uses the @2x file to display the high-res image. When the @2x file doesn't exist it takes the low-res image and scales it 2x which results in a blurry image.

Fair enough. In this case I assume that Apple has made sure that this naming convention does not create problems when used on different file systems...

basdelfos commented:
Hey, it's Apple. Assuming they made sure this naming convention doesn't create issues on other filesystems is not logical. But did a test on Windows 10 without any issue, ext3 and ext4 also do not have any issues with `@` in filenames.

mikeller commented:
Yeah, how true. We recently had a case at work where a subdirectory in the same directory had been created twice concurrently, once capitalised, once lower case. Everything was fine when checking out and building in the Windows shell (it's a .Net Framework project), since the two subdirectories got merged into one on the local disk, thanks to Windows' shitty file naming conventions.
However, if WSL was used to check out the project, all hell broke loose and the build failed with 'missing' source files that were actually in the other copy of the subdirectory...

basdelfos commented:
I plan to finish this PR this evening (including rebase) so this doesn't block any other PRs anymore.

basdelfos commented:
I think it is pretty much done for now.
- Moved all `*.js` files to the `js` folder.
- Moved all `*.css` files to the `css` folder.
- Moved `js/libraries` folder to root.
- Moved `changelog.html` to root.
- Moved `_locales` folder to root and changed name to `locales`.
- Moved `resources` folder to root.

Changed `gulpfile.js` to accommodate the above changes.

Tested on macOS and everything seems fine.

McGiverGim commented on gulpfile.js:
```diff
@@ -52,7 +52,7 @@ gulp.task('clean-release', clean_release);
 gulp.task('clean-cache', clean_cache);
 
-var distBuild = gulp.series(clean_dist, dist);
+var distBuild = gulp.series(clean_dist, dist_src, dist_locale, dist_libraries, dist_resources);
```
I don't know if it will be a good idea, but we can execute this tasks in parallel, if I'm not wrong:
>var distBuild = gulp.series(clean_dist, gulp.parallel(dist_src, dist_locale, dist_libraries, dist_resources));

basdelfos replied:
These tasks could run parallel to each other, the do not have conflicting actions at this point.
In my opinion there is less benefit, as the whole `dist` task runs in in less than 700ms.

McGiverGim commented:
Only for curiosity, one thing that I always asked to myself: why, in the `dist_src()` method we add the files one by one, in a lot of them, and not using the asterisk way to include all, maybe adding some exclusion if needed? I think is better to maintain, now when adding a `js` or `html` we must remember to modify the `gulpfile.js`.

I don't want to change it here, this can be done in a future PR if we see it has some sense.

mikeller commented:
Thanks @basdelfos for getting this done!

McGiverGim commented:
I think something is wrong, at least in my setup. It works, but if I make a: `npm install i18next --save`, the dependency is added to the `package.json` but it is not added to the `dist/node_modules` folder. Can someone test it?

McGiverGim commented:
Problem found, the:
>.pipe(install({
>            npm: '--production --ignore-scripts'
>        }));

is missing.

basdelfos commented:
@McGiverGim hmm, didn't know we used actual npm packages for running the app. Where is this package used for, as everything worked perfectly when I tested the app on my macbook?

McGiverGim commented:
No, it was for the "new" language selection in app. We are starting to use them, I had this PR ready: https://github.com/betaflight/betaflight-configurator/pull/879 but now it fails. No problem, I have modified the PR to add this again.

mikeller commented:
@basdelfos: Why was the installation of `npm` packages removed? Breaking the functionality of `npm` for no apparent reason does not make any sense to me. Maybe we should move away from using `npm` and use static libraries for everything if using `npm` is such a problem?

basdelfos commented:
@mikeller: there was no apparent reason why that code was in there as in the master - at that time - didn't use any static client-side libraries. I tested it and everything was working fine. What I didn't know was that there was another PR in the make that needed npm packages at the client-side that are included with npm. Didn't do it on purpose, this is something that can happen when more people work on the same code and someone can not track all the changes that are made all the time.

mikeller commented:
@basdelfos: The code was there to make our npm package adhere to the npm standard. Being standards compliant is reason enough in my opinion, as long as there is no negative impact. In additon to that, it enables tinkering with npm packages and testing if they should be used in the configurator, by making it easy to add them to the runtime environment.
